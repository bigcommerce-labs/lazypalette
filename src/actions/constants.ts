export const enum ConfigUpdateAction {
    PREVIEW = 'PREVIEW',
    SAVE = 'SAVE',
    PUBLISH =  'PUBLISH',
    UPDATE = 'UPDATE',
}

export const enum ToastMessages {
    ErrorSwitchToThemeEditor = 'There was an error switching to the old Theme Editor.',
    ErrorHistory = 'There was an error switching to the selected Theme History entry.',
    ErrorPreview = 'There was an error previewing your changes.',
    ErrorPublish = 'There was an error publishing your changes.',
    ErrorReset = 'There was an error restoring the original theme styles.',
    ErrorSave = 'There was an error saving your changes.',
    ErrorUpdate = 'There was an error applying your update.',
    ErrorVariation = 'There was an error switching styles.',
    Publish = 'Changes have been published.',
    Reset = 'Original theme styles restored.',
    Save = 'Changes have been saved.',
    Undo = 'Changes have been undone.',
    Update = 'Theme has been updated.',
}

export const enum ToastType {
    Success = 'Success',
    Error = 'Error',
}

export const enum ToastTimeout {
    Duration = 6000,
}
