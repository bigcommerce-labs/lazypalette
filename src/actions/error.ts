import { Action } from './action';

export enum ErrorActionTypes {
    CLEAR_ERRORS = 'CLEAR_ERRORS',
}

export interface ClearErrorsAction extends Action  {
    type: ErrorActionTypes.CLEAR_ERRORS;
}

export function clearErrors(): ClearErrorsAction {
    return {
        type: ErrorActionTypes.CLEAR_ERRORS,
    };
}
