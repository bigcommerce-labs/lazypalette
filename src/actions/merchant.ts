import { Dispatch } from 'redux';

import { State } from '../reducers/reducers';

export enum MerchantActionTypes {
    SET_STORE_DEFAULTS = 'SET_STORE_DEFAULTS',
    UPDATE_ACTIVE_THEME = 'UPDATE_ACTIVE_THEME',
}

export interface SetStoreDataAction {
    error?: boolean;
    payload: StoreDefaultData | Error;
    type: MerchantActionTypes.SET_STORE_DEFAULTS;
}

export interface StoreDefaultData {
    previewCode: string;
    activeThemeId: string;
    activeVariationId: string;
    isDownForMaintenance: boolean;
    isPrelaunchStore: boolean;
    storeHash: string;
    timezoneName: string;
    timezoneOffset: number;
}

export interface UpdateActiveThemeAction {
    payload: UpdateActiveThemeData;
    type: MerchantActionTypes.UPDATE_ACTIVE_THEME;
}

export interface UpdateActiveThemeData {
    activeThemeId: string;
    activeVariationId: string;
}

export function setStoreDefault(
    payload: StoreDefaultData | Error,
    error: boolean = false
): SetStoreDataAction {
    return {
        error,
        payload,
        type: MerchantActionTypes.SET_STORE_DEFAULTS,
    };
}

export function setStoreData(storeData: StoreDefaultData): Dispatch<State> {
    return (dispatch: Dispatch<State>) => dispatch(setStoreDefault(storeData));
}

export function updateActiveTheme(payload: UpdateActiveThemeData): UpdateActiveThemeAction {
    return {
        payload,
        type: MerchantActionTypes.UPDATE_ACTIVE_THEME,
    };
}
