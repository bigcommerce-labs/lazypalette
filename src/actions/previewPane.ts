import { entries } from 'lodash';
import { Dispatch } from 'redux';

import { ViewportType } from '../components/PreviewPane/PreviewPane';
import { State } from '../reducers/reducers';

import { Action } from './action';
import { ThemeConfigChange } from './theme';

export enum PreviewPaneActionTypes {
    PAGE_SOURCE_REQUEST = 'PAGE_SOURCE_REQUEST',
    PAGE_SOURCE_RESPONSE = 'PAGE_SOURCE_RESPONSE',
    PREVIEW_PANE_LOADED = 'PREVIEW_PANE_LOADED',
    PREVIEW_PANE_LOADING = 'PREVIEW_PANE_LOADING',
    THEME_FONT_CHANGE = 'THEME_FONT_CHANGE',
    THEME_PREVIEW_CONFIG_REQUEST = 'THEME_PREVIEW_CONFIG_REQUEST',
    VIEWPORT_CHANGE = 'VIEWPORT_CHANGE',
}

export interface PreviewPaneLoadingAction extends Action  {
    type: PreviewPaneActionTypes.PREVIEW_PANE_LOADING;
}
export function previewPaneLoading(): PreviewPaneLoadingAction {
    return {
        type: PreviewPaneActionTypes.PREVIEW_PANE_LOADING,
    };
}

export interface PreviewPaneLoadedAction extends Action  {
    type: PreviewPaneActionTypes.PREVIEW_PANE_LOADED;
}
export function previewPaneLoaded(): PreviewPaneLoadedAction {
    return {
        type: PreviewPaneActionTypes.PREVIEW_PANE_LOADED,
    };
}

export interface PageUrlRequest {
    page: string;
}
export interface PageUrlRequestAction extends Action  {
    payload: PageUrlRequest;
    type: PreviewPaneActionTypes.PAGE_SOURCE_REQUEST;
}
export function pageUrlRequest(payload: PageUrlRequest): PageUrlRequestAction {
    return {
        payload,
        type: PreviewPaneActionTypes.PAGE_SOURCE_REQUEST,
    };
}

export interface PageUrlResponse {
    page: string;
    pageUrl: string;
}
export interface PageUrlResponseAction extends Action  {
    error: boolean;
    payload: PageUrlResponse | Error;
    type: PreviewPaneActionTypes.PAGE_SOURCE_RESPONSE;
}
export function pageUrlResponse(
    payload: PageUrlResponse | Error,
    error: boolean = false
): PageUrlResponseAction {
    return {
        error,
        payload,
        type: PreviewPaneActionTypes.PAGE_SOURCE_RESPONSE,
    };
}

export interface FetchPageUrl {
    page: string;
}
export function fetchPageUrl(
    { page }: FetchPageUrl
) {
    return (dispatch: Dispatch<State>, getState: () => State) => {
        dispatch(pageUrlRequest({ page }));

        const { configurationId, lastCommitId, versionId } = getState().previewPane.themePreviewConfig;
        let pageSrc: string = page;

        if (versionId && configurationId) {
            const token = `${versionId}@${configurationId}`;
            const queryParams = {
                stencilEditor: lastCommitId ? `${token}@${lastCommitId}` : token,
            };
            const queryString = entries(queryParams).map(keyValuePair => keyValuePair.join('=')).join('&');

            pageSrc = page + (queryString ? `?${queryString}` : '');
        }

        return dispatch(pageUrlResponse({page, pageUrl: pageSrc }));
    };
}

export function updatePreviewPaneConfig(): (dispatch: Dispatch<State>, getState: () => State) => void {
    return (dispatch: Dispatch<State>, getState: () => State): void => {
        const {
            configurationId,
            variations,
            versionId,
        } = getState().theme;

        const isCurrentIndex = variations.map(variation => variation.isCurrent).indexOf(true);
        const lastCommitId = isCurrentIndex >= 0 ? variations[isCurrentIndex].lastCommitId : '';

        dispatch(previewPaneLoading());
        dispatch(receiveThemePreviewConfig({
            configurationId,
            lastCommitId,
            versionId,
        }));
    };
}

export interface RecieveThemePreviewConfig {
    configurationId: string;
    lastCommitId: string;
    versionId: string;
}
export interface RecieveThemePreviewConfigAction extends Action {
    payload: RecieveThemePreviewConfig;
    type: PreviewPaneActionTypes.THEME_PREVIEW_CONFIG_REQUEST;
}
export function receiveThemePreviewConfig(payload: RecieveThemePreviewConfig): RecieveThemePreviewConfigAction {
    return {
        payload,
        type: PreviewPaneActionTypes.THEME_PREVIEW_CONFIG_REQUEST,
    };
}

export function updateFonts(payload: ThemeConfigChange) {
    return {
        payload,
        type: PreviewPaneActionTypes.THEME_FONT_CHANGE,
    };
}

export interface ViewportChange {
    isRotated?: boolean;
    viewportType: ViewportType;
}
export interface ViewportChangeAction extends Action {
    payload: ViewportChange;
    type: PreviewPaneActionTypes.VIEWPORT_CHANGE;
}
export function viewportChange(payload: ViewportChange): ViewportChangeAction {
    return {
        payload,
        type: PreviewPaneActionTypes.VIEWPORT_CHANGE,
    };
}
