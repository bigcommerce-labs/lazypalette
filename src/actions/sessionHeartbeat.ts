import { SessionHeartbeatResponse } from '../services/sessionHeartbeat';

export enum SessionHeartbeatActionTypes {
    SESSION_HEARTBEAT_RESPONSE = 'SESSION_HEARTBEAT_RESPONSE',
}

export interface SessionHeartbeatResponseAction {
    error: boolean;
    payload: SessionHeartbeatResponse | Error;
    type: SessionHeartbeatActionTypes.SESSION_HEARTBEAT_RESPONSE;
}

export function sessionHeartbeatResponse(
    payload: SessionHeartbeatResponse,
    err: Error | null
): SessionHeartbeatResponseAction {
    if (err) {
        return {
            error: true,
            payload: err,
            type: SessionHeartbeatActionTypes.SESSION_HEARTBEAT_RESPONSE,
        };
    }

    return {
        error: false,
        payload,
        type: SessionHeartbeatActionTypes.SESSION_HEARTBEAT_RESPONSE,
    };
}
