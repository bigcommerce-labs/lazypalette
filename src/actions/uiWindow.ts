import { ColorResult } from 'react-color';
import uuid from 'uuid';

export enum UIWindowActionTypes {
    OPEN_UI_WINDOW = 'OPEN_UI_WINDOW',
    CLOSE_UI_WINDOW = 'CLOSE_UI_WINDOW',
}

export enum UIWindowTypes {
    COLOR_PICKER = 'COLOR_PICKER',
}

export interface UIWindowContent {
    position: { x: number, y: number };
}

export interface ColorPickerContent extends UIWindowContent {
    color: string;
    onChange?(value: ColorResult): void;
}

export interface OpenUIWindowAction<Content> {
    payload: {
        content: Content;
        id: string;
        type: UIWindowTypes;
    };
    type: UIWindowActionTypes.OPEN_UI_WINDOW;
}

export interface CloseUIWindowAction {
    payload: {
        id: string,
    };
    type: UIWindowActionTypes.CLOSE_UI_WINDOW;
}

export interface OpenColorPickerArgs {
    color: string;
    id?: string;
    position: { x: number, y: number };
    onChange?(value: ColorResult): void;
}

export function openColorPicker(content: OpenColorPickerArgs): OpenUIWindowAction<ColorPickerContent> {
    return {
        payload: {
            content: {
                color: content.color,
                onChange: content.onChange,
                position: content.position,
            },
            id: content.id || uuid(),
            type: UIWindowTypes.COLOR_PICKER,
        },
        type: UIWindowActionTypes.OPEN_UI_WINDOW,
    };
}

export function closeUIWindow(id: string): CloseUIWindowAction {
    return {
        payload: { id },
        type: UIWindowActionTypes.CLOSE_UI_WINDOW,
    };
}
