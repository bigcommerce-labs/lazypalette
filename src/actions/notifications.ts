import { Dispatch } from 'redux';

import { State } from '../reducers/reducers';

import { Action } from './action';
import { ToastTimeout } from './constants';

export enum NotificationActionTypes {
    OPEN_NOTIFICATION = 'OPEN_NOTIFICATION',
    CLOSE_NOTIFICATION = 'CLOSE_NOTIFICATION',
}

export interface OpenNotificationAction extends Action  {
    payload: NotificationsProps;
    type: NotificationActionTypes.OPEN_NOTIFICATION;
}

export interface CloseNotificationAction extends Action  {
    type: NotificationActionTypes.CLOSE_NOTIFICATION;
}

export interface NotificationsProps {
    autoDismiss: boolean;
    message: string;
    type: string;
}

export function openNotification(content: NotificationsProps): OpenNotificationAction {
    return {
        payload: {
            autoDismiss: content.autoDismiss,
            message: content.message,
            type: content.type,
        },
        type: NotificationActionTypes.OPEN_NOTIFICATION,
    };
}

export function closeNotification(): CloseNotificationAction {
    return {
        type: NotificationActionTypes.CLOSE_NOTIFICATION,
    };
}

export function createNotification(autoDismiss: boolean, message: string, type: string) {
    return (dispatch: Dispatch<State>) => {
        dispatch(openNotification({ autoDismiss, message, type }));

        if (autoDismiss) {
            setTimeout(() => {
                dispatch(closeNotification());
            }, ToastTimeout.Duration);
        }
    };
}
