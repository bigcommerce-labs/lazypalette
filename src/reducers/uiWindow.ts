import { Action } from '../actions/action';
import { UIWindowActionTypes, UIWindowContent, UIWindowTypes } from '../actions/uiWindow';

export interface UIWindowData {
    content: UIWindowContent;
    id: string;
    type: UIWindowTypes;
}

export interface UIWindowState {
    uiWindows: UIWindowData[];
}

const initialState: UIWindowState = {
    uiWindows: [],
};

function uiWindow(state: UIWindowState = initialState, action: Action): UIWindowState {
    if (action.error) {
        return state;
    }

    switch (action.type) {
        case UIWindowActionTypes.OPEN_UI_WINDOW:
            return { ...state,
                uiWindows: [...state.uiWindows,
                    {
                        content: action.payload.content,
                        id: action.payload.id,
                        type: action.payload.type,
                    }],
            };
        case UIWindowActionTypes.CLOSE_UI_WINDOW:
            return { ...state, uiWindows: state.uiWindows.filter(m => m.id !== action.payload.id) };
        default:
            return state;
    }
}

export default uiWindow;
