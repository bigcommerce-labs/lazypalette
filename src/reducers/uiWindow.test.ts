import * as uiWindowActions from '../actions/uiWindow';

import uiWindow, { UIWindowState } from './uiWindow';

const initialState: UIWindowState = {
    uiWindows: [
        {
            content: {
                position: { x: 5, y: 10 },
            },
            id: '5678',
            type: uiWindowActions.UIWindowTypes.COLOR_PICKER,
        },
        {
            content: {
                position: { x: 15, y: 20 },
            },
            id: '0987',
            type: uiWindowActions.UIWindowTypes.COLOR_PICKER,
        },
    ],
};

describe('uiWindow reducer', () => {
    describe('when we dispatch an openColorPicker action', () => {
        it('adds a window to its state', () => {
            const action = uiWindowActions.openColorPicker({
                color: '#FFFFFF',
                id: '1234',
                position: { x: 10, y: 10 },
            });

            expect(uiWindow(initialState, action)).toMatchSnapshot();
        });
    });

    describe('when we dispatch an closeUIWindow action', () => {
        it('removes a window from its state', () => {
            const action = uiWindowActions.closeUIWindow('5678');

            expect(uiWindow(initialState, action)).toMatchSnapshot();
        });
    });

    describe('when we dispatch an error action', () => {
        it('does not change its state', () => {
            const action = uiWindowActions.closeUIWindow('5678');
            action['error'] = true; // tslint:disable-line

            expect(uiWindow(initialState, action)).toMatchSnapshot();
        });
    });

    describe('when we dispatch an unrelated action', () => {
        it('does not change its state', () => {
            const action = {
                type: 'fred',
            };

            expect(uiWindow(initialState, action)).toMatchSnapshot();
        });
    });
});
