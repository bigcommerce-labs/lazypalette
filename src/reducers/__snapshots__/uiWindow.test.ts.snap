// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uiWindow reducer when we dispatch an closeUIWindow action removes a window from its state 1`] = `
Object {
  "uiWindows": Array [
    Object {
      "content": Object {
        "position": Object {
          "x": 15,
          "y": 20,
        },
      },
      "id": "0987",
      "type": "COLOR_PICKER",
    },
  ],
}
`;

exports[`uiWindow reducer when we dispatch an error action does not change its state 1`] = `
Object {
  "uiWindows": Array [
    Object {
      "content": Object {
        "position": Object {
          "x": 5,
          "y": 10,
        },
      },
      "id": "5678",
      "type": "COLOR_PICKER",
    },
    Object {
      "content": Object {
        "position": Object {
          "x": 15,
          "y": 20,
        },
      },
      "id": "0987",
      "type": "COLOR_PICKER",
    },
  ],
}
`;

exports[`uiWindow reducer when we dispatch an openColorPicker action adds a window to its state 1`] = `
Object {
  "uiWindows": Array [
    Object {
      "content": Object {
        "position": Object {
          "x": 5,
          "y": 10,
        },
      },
      "id": "5678",
      "type": "COLOR_PICKER",
    },
    Object {
      "content": Object {
        "position": Object {
          "x": 15,
          "y": 20,
        },
      },
      "id": "0987",
      "type": "COLOR_PICKER",
    },
    Object {
      "content": Object {
        "color": "#FFFFFF",
        "onChange": undefined,
        "onClose": undefined,
        "position": Object {
          "x": 10,
          "y": 10,
        },
      },
      "id": "1234",
      "type": "COLOR_PICKER",
    },
  ],
}
`;

exports[`uiWindow reducer when we dispatch an unrelated action does not change its state 1`] = `
Object {
  "uiWindows": Array [
    Object {
      "content": Object {
        "position": Object {
          "x": 5,
          "y": 10,
        },
      },
      "id": "5678",
      "type": "COLOR_PICKER",
    },
    Object {
      "content": Object {
        "position": Object {
          "x": 15,
          "y": 20,
        },
      },
      "id": "0987",
      "type": "COLOR_PICKER",
    },
  ],
}
`;
